src/client.rs

```diff
@@ -12,48 +16,67 @@ pub enum Config {
     },
 }
 
-pub struct EjudgeClient {
-    pub session_id: String,
-    pub base_url: url::Url,
-    pub client: reqwest::Client,
+#[derive(Debug, StructOpt)]
+pub struct Submission {
+    #[structopt(short, long)]
+    pub problem_id: Option<String>,
+
+    #[structopt(short, long, default_value = "3")]
+    pub lang_id: String,
+
+    #[structopt(parse(from_os_str))]
+    pub file: PathBuf,
+}
+
+fn get_problem_id_from_filename(name: &std::ffi::OsStr) -> Option<String> {
+    if name.len() == 1 {
+        let order = match name.to_str()?.chars().next().unwrap() {
+            x @ 'A'..='Z' => x as u32 - 'A' as u32 + 1,
+            x @ 'a'..='z' => x as u32 - 'a' as u32 + 1,
+            _ => {
+                return None;
+            }
+        };
+
+        Some(format!("{}", order))
+    } else {
+        None
+    }
 }
 
-impl EjudgeClient {
-    pub fn from_config(config: Config) -> Result<EjudgeClient> {
-        match config {
-            Config::Ejudge {
-                contest_url,
-                session_id,
-            } => Ok(EjudgeClient {
-                session_id: session_id.clone(),
-                base_url: url::Url::parse(&contest_url)?,
-                client: reqwest::Client::builder().cookie_store(true).build()?,
-            }),
-        }
+pub fn get_problem_id(submission: &Submission) -> Result<String> {
+    if let Some(pid) = &submission.problem_id {
+        Ok(pid.clone())
+    } else {
+        [&std::env::current_dir()?, &submission.file]
+            .iter()
+            .filter_map(|path| path.file_stem())
+            .filter_map(|name| get_problem_id_from_filename(&name))
+            .next()
+            .ok_or(Error::MissingProblemId)
     }
+}
 
-    pub fn from_env() -> Result<EjudgeClient> {
+#[async_trait]
+pub trait Client: Sized {
+    /// Builds client with given config.
+    fn from_config(config: Config) -> Result<Self>;
```

Comment:
Why `fn from_config(config: Config) -> Result<Self>` appears here? For reference, `enum Config` designed to have several variants:
```rust
pub enum Config {
    Ejudge {
        contest_url: String,
        session_id: String,
    },
}
```
What happens if someone tries to build `CodeforcesClient` with `Config::Ejudge{...}`? Probably, functions `from_config` as well as `from_env` should've existed outside `trait Client` and return something like 
```rust
pub fn from_config(config: Config) -> Result<Box<dyn Client>>
```

To keep things consistent you may introduce 
```rust
trait Client {
  type Config; 
  
  fn from_config(config: Config) -> Result<Self>;
  fn get_config(&self) -> Config;
  
  fn save_config(path: PathBuf) -> Result<()> {
     // default implementation goes here
  }
}
```

Reply:
Already done with basically two traits: `SubmitClient` and `ConfigClient`: the first one is capable of submitting, the second is about storing/loading its state from file.